Fetch sensei extensions from sensei site

renatho opened the pull request:
Fixes #3044

### Changes proposed in this Pull Request

* Fetch the extensions from `https://senseilms.com/wp-json/senseilms-products/1.0/search?category=setup-wizard-extensions` - I created this new category `setup-wizard-extensions` to add only the extensions that we want to appear as suggestion.

### Testing instructions

* Go to `/wp-admin/admin.php?page=sensei_onboarding&step=features` and check if the extensions are being listed.

### Some questions

* Should we really override the titles? Or could we rename it in the `senseilms.com`?
* Do we need to worry about the [dynamic data](https://senseilms.com/wp-json/senseilms-products/1.0/search?category=setup-wizard-extensions) translations and currency conversion?
* Are good the links coming from the JSON? They are pointing to senseilms.com and redirecting to WC.com or WP.org.
* About the sorting, do we have a specific rule to apply in the array? I had thought about creating an array with the sort, but if we add more extensions in the future it would not appear in the order that we want.

renatho commented:
@donnapep, this PR is not ready yet, waiting for other PR. But if you could check the questions that I put in the PR description, I could address these things until the other PR is merged.

donnapep commented:
> * Should we really override the titles? Or could we rename it in the `senseilms.com`?

There's no need to override these. The design isn't accurate. We can use whatever is coming from the API.

> * Do we need to worry about the [dynamic data](https://senseilms.com/wp-json/senseilms-products/1.0/search?category=setup-wizard-extensions) translations and currency conversion?

Do you mean should we make the strings translatable? If so, then yes.

I don't think we need to worry about converting currencies, at least not for a first iteration.

> * Are good the links coming from the JSON? They are pointing to senseilms.com and redirecting to WC.com or WP.org.

This is fine. We have Google Analytics set up on SenseiLMS.com and when users click a link in Sensei LMS > Extensions, it enables us to capture them on SenseiLMS.com before they head to WC.com.

> * About the sorting, do we have a specific rule to apply in the array? I had thought about creating an array with the sort, but if we add more extensions in the future it would not appear in the order that we want.

Let's show them in the order in which they're being returned. If it becomes important in the future, perhaps we can consider using WC's _Menu Order_ product setting.

yscik commented on includes/admin/class-sensei-onboarding.php:
```diff
@@ -385,4 +386,112 @@ public function update_wizard_user_data( $changes ) {
+		$json = json_decode( $raw_extensions['body'], true );
+
+		// Normalize property names.
+		$properties = [
+			'product_slug' => 'id',
+			'excerpt'      => 'description',
+			'link'         => 'learnMoreLink',
```
Instead of changing the property names, WDYT about changing them on the frontend to use the original names? Should be simpler if it's not too much work.

renatho replied:
Looks good! I just added the `slug` prop, so we don't need to normalize it in the frontend, because using with `_` doesn't match with our lint rules. https://github.com/Automattic/sensei/pull/3176/commits/e1ba40aae4777e79414ecdc024f73a4432dde6d0

renatho replied:
I refactored again to normalize it in the frontend, because I had to normalize also the feature titles ;)
https://github.com/Automattic/sensei/pull/3176/commits/42e54d6851e7638f2bf0dabc7a4581a7d29e73a9
https://github.com/Automattic/sensei/pull/3176/commits/5b56e6613f3acb21314722ad3a2cb04dd159e61b

yscik commented on includes/admin/class-sensei-onboarding.php:
```diff
@@ -385,4 +386,112 @@ public function update_wizard_user_data( $changes ) {
+	 * @return array Sensei extensions.
+	 */
+	public function get_sensei_extensions() {
+		$extensions = get_transient( self::EXTENSIONS_TRANSIENT );
+
+		if ( false === $extensions ) {
+			$data = wp_remote_get( 'https://senseilms.com/wp-json/senseilms-products/1.0/search?category=setup-wizard-extensions' );
+			if ( is_wp_error( $data ) ) {
```
There is an existing method `Sensei_Extensions::get_extensions` that this mostly duplicates. Should we make that public (can even be static) and use it here too? It can accept a category parameter and caches the result for it, and also handles the JSON body decoding.

renatho replied:
Ah, great catch! Refactored to use the same method here: https://github.com/Automattic/sensei/pull/3176/commits/817af2fb16499b9d656e43ee8ef4b4865d9d6543

yscik commented:
Should adding the real installed status of the plugins be a part of this PR? 

Looks like `get_plugins` is not loaded by default for REST API calls, so something like this will be needed for that:
https://developer.wordpress.org/reference/functions/get_plugins/#comment-1207

renatho commented:
> Should adding the real installed status of the plugins be a part of this PR?
> 
> Looks like `get_plugins` is not loaded by default for REST API calls, so something like this will be needed for that:
> https://developer.wordpress.org/reference/functions/get_plugins/#comment-1207

Thank you for the suggestion! My idea is to do it in this issue: https://github.com/Automattic/sensei/issues/3095. If I remember well, I checked in the WC code that they do a validation checking if the plugin file exists. But I'll jot down your suggestion because it can be useful =)

yscik commented on assets/onboarding/data/actions.js:
```diff
@@ -37,9 +39,7 @@ export function* fetchSetupWizardData() {
-		yield successFetch( {
-			...data,
-		} );
+		yield successFetch( normalizeSetupWizardData( data ) );
```
Should this be done in a reducer instead? Seems like a better fit for transforming data.

renatho replied:
I usually like to do this type of thing as early as possible. In the action, in this case. Because if we want to get the same action in other reducers, or create a middleware to handle these data, this is ready. Make sense to you?

yscik replied:
Yeah, I can accept this thinking too :)

yscik commented on assets/onboarding/data/normalizer.js:
```diff
@@ -0,0 +1,47 @@
+ * @param {string|number} price    Feature price.
+ * @param {string}        [status] Feature status.
+ */
+const addDetailsToTitle = ( title, price, status ) => {
```
Just a minor improvement idea, this could be called like `addDetailsToTitle( feature )` with a signature of ` ( { title, price, status } ) => { ... }`.

WDYT about naming it something like `getTitleWithDetails` - `add` feels like it would modify something.

renatho replied:
Yeah! It's better! Refactored here: https://github.com/Automattic/sensei/pull/3176/commits/dbb97711700668e23902e58caf454c069856caef

yscik commented on includes/rest-api/class-sensei-rest-api-setup-wizard-controller.php:
```diff
@@ -195,6 +195,7 @@ public function get_data() {
 			],
 			'features'        => [
 				'selected' => $user_data['features'],
+				'options'  => $this->setup_wizard->get_sensei_extensions(),
```
The new `options` should be added to the `features` object in `get_schema` (which also has a old `plugins` property now that can be removed).

renatho replied:
Good catch! Fixed here: https://github.com/Automattic/sensei/pull/3176/commits/70a52d75ae492eeb0e5b90af315dee8a88f04412

renatho commented:
I added a filter to prevent PHP requests running tests. https://github.com/Automattic/sensei/pull/3176/commits/db5678c4a2da68c39612895d9a6a478f26f3be20

If we want to test a request, we can override this filter. Different suggestions are welcome.

yscik commented:
> I added a filter to prevent PHP requests running tests. [db5678c](https://github.com/Automattic/sensei/commit/db5678c4a2da68c39612895d9a6a478f26f3be20)

Not sure if it would be a good idea, but we could have it throw an error by default, so that it has to be overwritten with an explicit return value or handler when testing code making external requests. Getting an empty array from requests could lead to some confusion in a few weeks when we forget about this.

renatho commented:
> > I added a filter to prevent PHP requests running tests. [db5678c](https://github.com/Automattic/sensei/commit/db5678c4a2da68c39612895d9a6a478f26f3be20)
> 
> Not sure if it would be a good idea, but we could have it throw an error by default, so that it has to be overwritten with an explicit return value or handler when testing code making external requests. Getting an empty array from requests could lead to some confusion in a few weeks when we forget about this.

Hum! Good idea! I added the filter to the API tests where it was broken and added a throw in a filter with priority 99 checking whether it was filtered or not. https://github.com/Automattic/sensei/pull/3176/commits/ccf3617d9487883a1fa5398cbe828631fdcadd64

donnapep commented:
@renatho FYI that we don't want to show WCPC on this list as that needs to be installed from WC.com, which we are not doing in this iteration.
